domain/secretbackend/state/operation.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/secretbackend"
+	backenderrors "github.com/juju/juju/domain/secretbackend/errors"
+	"github.com/juju/juju/internal/database"
+)
+
+type upsertOperation struct {
```

Comment:
I don't see what this gains us.

It is a new type for a single use, with validation, prepare and execution phases that could just be in the state method. It just adds opacity.